juntagrico_billing/util/pdfbill.py

```diff
@@ -0,0 +1,253 @@
+import datetime
+from reportlab.platypus import SimpleDocTemplate, Paragraph, Spacer, Table
+from reportlab.lib.styles import getSampleStyleSheet, ParagraphStyle
+from reportlab.lib.units import cm
+from django.utils.formats import get_format
+from django.utils.dateformat import format
+from django.utils.translation import gettext as _
+from juntagrico.config import Config
+from juntagrico_billing.entity.settings import Settings
+from juntagrico_billing.util.qrbill import get_qrbill_svg, is_qr_iban
+from svglib.svglib import SvgRenderer
+from lxml import etree
+
+
+class PdfBillRenderer(object):
+
+    def __init__(self):
+        #
+        # define styles
+        #
+        self.org_heading = getSampleStyleSheet()['Heading1']
+
+        self.normal = ParagraphStyle(name='Normal')
+        self.normalright = ParagraphStyle(
+                            'Normal-Right',
+                            parent=self.normal, alignment=2)
+        self.text = ParagraphStyle(
+                            name='Text',
+                            parent=self.normal, spaceBefore=6)
+        self.heading1 = ParagraphStyle(
+                            'Heading1',
+                            parent=self.normal,
+                            fontName=self.org_heading.fontName,
+                            fontSize=14, leading=18,
+                            spaceBefore=12, spaceAfter=6)
+        self.heading2 = ParagraphStyle(
+                            'Heading2', parent=self.normal,
+                            fontName=self.org_heading.fontName,
+                            spaceBefore=6, spaceAfter=2)
+
+        self.table_style = [
+            ('ALIGN', (0, 0), (-1, -1), 'LEFT'),    # left align all cells
+            ('VALIGN', (0, 0), (-1, -1), 'TOP'),    # valign top all cells
+            ('TOPPADDING', (0, 0), (-1, -1), 1),    # reduce top and bottom
+                                                    # padding (default is 3)
+            ('BOTTOMPADDING', (0, 0), (-1, -1), 1),
+            ('LEFTPADDING', (0, 0), (0, -1), 0)     # right align last row
+        ]
+
+    def render(self, bill, outfile):
+        """
+        render a bill as PDF in to a byte file-like output.
+        """
+        story = []
+
+        story.append(Spacer(1, 2*cm))
+        self.render_addresses(bill, story)
+        story.append(Spacer(1, 3*cm))
+        self.render_header(bill, story)
+        story.append(Spacer(1, 0.5*cm))
+        self.render_bill_items(bill, story)
+
+        # public notes
+        if bill.public_notes:
+            story.append(Paragraph(bill.public_notes, self.text))
+
+        self.render_payments(bill, story)
+
+        # payment notice
+        # todo: make setting for due date notice link
+        story.append(Paragraph('%s %s.' % (
+            _('The billed amount is due for payment according to ' +
+              'the due date notice found here:'),
+            'https://ortoloco.ch/dokumente/FäHi_2021.pdf'), self.text))
```

Comment:
Setting

Reply:
Habe setting DUEDATE_NOTICE_URL eingeführt. Verwendet in user_bill.html, pdfbill und bill_notification.txt.